jupyter_server_fileid/manager.py

```diff
@@ -229,26 +346,43 @@ def _validate_root_dir(self, proposal):
         normalized_content_root = self._normalize_separators(proposal["value"])
         return normalized_content_root
 
+    def _create_tables(self):
+        self.con.execute(
+            "CREATE TABLE IF NOT EXISTS Files("
+            "id TEXT PRIMARY KEY NOT NULL, "
+            "path TEXT NOT NULL UNIQUE"
+            ")"
+        )
+        self.con.execute("CREATE INDEX IF NOT EXISTS ix_Files_path ON Files (path)")
+
     def __init__(self, *args, **kwargs):
         # pass args and kwargs to parent Configurable
         super().__init__(*args, **kwargs)
+
         # initialize instance attrs
         self._update_cursor = False
+
         # initialize connection with db
         self.log.info(f"ArbitraryFileIdManager : Configured root dir: {self.root_dir}")
         self.log.info(f"ArbitraryFileIdManager : Configured database path: {self.db_path}")
+
         self.con = sqlite3.connect(self.db_path)
-        self.log.info("ArbitraryFileIdManager : Successfully connected to database file.")
-        self.log.info("ArbitraryFileIdManager : Creating File ID tables and indices.")
         # do not allow reads to block writes. required when using multiple processes
         self.con.execute("PRAGMA journal_mode = WAL")
-        self.con.execute(
-            "CREATE TABLE IF NOT EXISTS Files("
-            "id TEXT PRIMARY KEY NOT NULL, "
-            "path TEXT NOT NULL UNIQUE"
-            ")"
-        )
-        self.con.execute("CREATE INDEX IF NOT EXISTS ix_Files_path ON Files (path)")
+
+        self.log.info("ArbitraryFileIdManager : Successfully connected to database file.")
+
+        if self._check_manifest():
+            # ensure tables and indices are created
+            self.log.info("ArbitraryFileIdManager : Creating File ID tables and indices.")
+            self._create_tables()
+        else:
+            # migrate database from old manager to current
+            self.log.info(
+                "ArbitraryFileIdManager : Database from different File ID manager detected. Migrating tables and indices."
+            )
+            self._migrate_database()
```

Comment:
Could this lead to thrashing?  For example, I use two classes of FIDs (e.g., LFID and AFID) configured on my system, neither of which have bothered to configure the db path (plausible if the LFID created the Files table since its schema is a superset of AFID), then won't the AFID trigger migration when it runs, and vice versa when the LFID starts?

I really think this kind of functionality should be in the user's face and performed by a CLI tool that also does `import` (from a source), `export` (to a destination), and `migration` using both.   With a CLI tool, you can also support _any_ database type or storage type and no need for a `Manifest` table.  In addition, you could support migration between databases of the same type (e.g., SQLite).

If you're too strict about class names in the manifest, you could also run into issues where subclasses are treated as distinct classes, despite the fact that their schemas (and base functionality) are identical.

Reply:
Yes, the migration is being done automatically at the present moment.

> I really think this kind of functionality should be in the user's face and performed by a CLI tool that also does import (from a source), export (to a destination), and migration using both. With a CLI tool, you can also support any database type or storage type and no need for a Manifest table. In addition, you could support migration between databases of the same type (e.g., SQLite).

Wait, but what would `import` and `export` do? These CLI options don't make sense to me. Importing requires somewhere to export to, and vice versa.

I do agree that migrations should be made explicit and only done via a CLI tool. This allows us to convey the backup database path to the end user more easily. However, I'm wary of the old behavior this would lead to:

1. start server with ArbitraryFIDM
2. stop server and restart with LocalFIDM
3. File ID extension fails to load, but this can get easily lost in the logs
4. all extensions depending on File ID are now broken and we could lose data since we're no longer able to index files

Do we know of a way to stop the server from starting if the server extension fails? That way, we can force the user to either migrate their database or switch back to their old FIDM.

Reply:
> Wait, but what would import and export do? These CLI options don't make sense to me. Importing requires somewhere to export to, and vice versa.

Correct.  Import and Export need sources and destinations and it would be up to the CLI tool to define the types of sources (FIDMs, csv files, jsons, etc.).  _Migration_ essentially reverses those, where the _source_ is exported from and the _destination_ is imported to and would minimally support FIDM "source types".  I.e., the initial implementation of the tool would take FIDM classes from which instances are instantiated and used.  Support for other types could be added as necessary.

Regarding the failure scenario bullets...
Why does the FileID service fail to load when switching to the LFIDM - is it because the schema from the AFIDM is not sufficient and they're both configured to use the same database file (which strikes me as an administration bug)?
Won't _every_ call to the FID service then fail and wouldn't that only go unnoticed if the calling application ignored exceptions?  I hope callers are robust enough to recognize exceptions, but I might be missing something here.

Reply:
> Correct. Import and Export need sources and destinations and it would be up to the CLI tool to define the types of sources (FIDMs, csv files, jsons, etc.).

Sorry if I'm not understanding you fully, but it almost sounds like you want two additional CLI options in addition to `jupyter server fileid migrate`:

- `jupyter server fileid import`
- `jupyter server fileid export`

My question is what would these options do differently than `migrate`? It seems importing a file ID database and exporting to a different one is the exact same as migrate.

> Why does the FileID service fail to load when switching to the LFIDM - is it because the schema from the AFIDM is not sufficient and they're both configured to use the same database file (which strikes me as an administration bug)?

Yeah, so it'll fail at runtime because the database schemas are different. While this is partially an administration bug, there is a legitimate use case for migrations regardless. Right now, if you want to switch FIDM implementations, you're forced to use a separate DB file, meaning you just lost all the file IDs you previously indexed. Thus you also broke all extensions that depend on file ID.

> Won't every call to the FID service then fail and wouldn't that only go unnoticed if the calling application ignored exceptions? I hope callers are robust enough to recognize exceptions, but I might be missing something here.

Right, but there remains the issue of how to communicate to the user that "because you switched file ID managers, the best way to fix these errors is to migrate your database". I would prefer the server to just not start if the file ID database hasn't been migrated rather than have the user try and debug this on their own. That way, we force the user to take one of three paths:

1. Revert back to the old File ID manager
2. Perform a migration with the CLI tool
3. Drop the old database

I don't want people to have to scratch their heads and figure out why file ID isn't working after switching implementations.

Reply:
> Sorry if I'm not understanding you fully, but it almost sounds like you want two additional CLI options

I'm just saying that since _migration_ is the "pipe between export and import" a CLI tool could offer different sources.  What we've been primarily talking about are sources of FIDM instances (since they front the database), but I'm saying that you could also support different _output destinations_ where the CLI can store exported data, and different _input sources_ where the CLI tool can get data for importing.

So I could perform an equivalent migration by exporting from FIDM1 to a file `exported.json` and importing from `exported.json` to FIDM2.  That is, the CLI tool manages the "pipe" since it is the caller of `FIDM1.export_rows()` and `FIDM2.import_rows()`.  This can then provide users the ability to have more choices for how they want to migrate (or _backup_) their FID databases.

One application for this may be that the user needs to perform some transformations on the exported rows (e.g., `exported.json`) prior to their import as part of their "migration plan".

We don't necessarily need to expose other sources and split `migrate` into `export` and `import` right away but should have that kind of thing in mind during implementation so that "the pipe" is extensible such that a user could manually intervene between "export" and "import" as part of their migration.

Reply:
Thinking a bit more on this, the first non-FIDM source type to support would probably want to be `csv`, since you want the ability to read _rows_ from a file and JSON doesn't really lend itself to that format (that I'm aware of).  `.csv` would also be immediately usable in spreadsheet applications in which the transformations could be performed, etc.

Reply:
I don't think we need to support import/export to some secondary file format in this PR. Let's limit the scope of this PR for just migrating between SQLite DBs. However, these are great ideas we should definitely think about implementing later down the road.